File: 2448-count-number-of-bad-pairs/count-number-of-bad-pairs.cpp
class Solution {
public:
    long long countBadPairs(vector<int>& nums) {
        int n=nums.size();
        unordered_map<long long, int> mp;
        for(int i=0;i<nums.size();i++){
            mp[i-nums[i]]++;
        }
        long long ans=0;
        for(auto it:mp){
            while(it.second--){
                n=n-1;
                ans=ans+n-it.second;
            }
            if(it.second==0) mp.erase(it.first);
        }
        return ans; 
    }
};

File: 2448-count-number-of-bad-pairs/count-number-of-bad-pairs.java
class Solution {
    long sum(long n){
        return (long)n*(long)(2+(n-1))/2;
    }
    public long countBadPairs(int[] nums) {
        int n=nums.length;
        HashMap<Long,Long>map=new HashMap<>();
        long pairs=(sum(n-1));
        for(int i=0;i<n;i++){
            long diff=(long)nums[i]-i;
            if(map.containsKey(diff)){
                map.put(diff,map.get(diff)+1);
            }
            else{
                map.put(diff,(long)1);
            }
        }
        long sum1=0;
        for(long ele:map.values()){
            if(ele>1){
                sum1+=sum(ele-1);
            }
        }
        return pairs-sum1;
    }
}